ftl/node/builder.py

```diff
@@ -0,0 +1,99 @@
+# Copyright 2017 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""This package defines the interface for orchestrating image builds."""
+
+import hashlib
+import os
+import subprocess
+import tempfile
+
+
+from containerregistry.client.v2_2 import append
+from ftl.common import builder
+
+
+_NODE_NAMESPACE = 'node-package-lock-cache'
+_PACKAGE_LOCK = 'package-lock.json'
+_PACKAGE_JSON = 'package.json'
+
+
+class Node(builder.JustApp):
+
+    def __init__(self, ctx):
+        super(Node, self).__init__(ctx)
+
+    def __enter__(self):
+        """Override."""
+        return self
+
+    def CreatePackageBase(self, base_image, cache):
+        """Override."""
+        if self._ctx.Contains(_PACKAGE_LOCK):
+            print('Using %s as a package descriptor.' % _PACKAGE_LOCK)
+            descriptor = self._ctx.GetFile(_PACKAGE_LOCK)
+        elif self._ctx.Contains(_PACKAGE_JSON):
+            print('Using %s as a package descriptor.' % _PACKAGE_JSON)
+            descriptor = self._ctx.GetFile(_PACKAGE_JSON)
+        else:
+            print('No package descriptor found. Not installing anything.')
+            return base_image
```

Comment:
Why not use a loop here (as you have below)?

Why not record the filename here so that we avoid the loop below?